kernel/patches/4.19.x/0001-enable-ptp_kvm-for-arm64-on-4.19.patch

```diff
@@ -0,0 +1,421 @@
++         * ts2 is the time measured exactly after the hvc call back;
++         * the half of the diffrence between ts2 and ts1 is the
++         * time compensation.
++         */
++        
++        *device_time = timespec64_to_ktime(ts64);
++        system_counter->cycles = hvc_res.a2;
++        spin_unlock(&kvm_ptp_lock);
++        system_counter->cs = &clocksource_counter;
++        return 0;
+ }
++#endif
++
++#ifdef CONFIG_X86
++static int ptp_kvm_get_time_fn(ktime_t *device_time,
++                               struct system_counterval_t *system_counter,
++                               void *ctx)
++{
++        unsigned long ret;
++        struct timespec64 tspec;
++        unsigned version;
++        int cpu;
++        struct pvclock_vcpu_time_info *src;
++
++        spin_lock(&kvm_ptp_lock);
++
++        preempt_disable_notrace();
++        cpu = smp_processor_id();
++        src = &hv_clock[cpu].pvti;
++
++        do {
++                /*
++                 * We are using a TSC value read in the hosts
++                 * kvm_hc_clock_pairing handling.
++                 * So any changes to tsc_to_system_mul
++                 * and tsc_shift or any other pvclock
++                 * data invalidate that measurement.
++                 */
++                version = pvclock_read_begin(src);
++
++                ret = kvm_hypercall2(KVM_HC_CLOCK_PAIRING,
++                                     clock_pair_gpa,
++                                     KVM_CLOCK_PAIRING_WALLCLOCK);
++                if (ret != 0) {
++                        pr_err_ratelimited("clock pairing hypercall ret %lu\n", ret);
++                        spin_unlock(&kvm_ptp_lock);
++                        preempt_enable_notrace();
++                        return -EOPNOTSUPP;
++                }
++
++                tspec.tv_sec = clock_pair.sec;
++                tspec.tv_nsec = clock_pair.nsec;
++                ret = __pvclock_read_cycles(src, clock_pair.tsc);
++        } while (pvclock_read_retry(src, version));
++
++        preempt_enable_notrace();
++
++        system_counter->cycles = ret;
++        system_counter->cs = &kvm_clock;
++
++        *device_time = timespec64_to_ktime(tspec);
++
++        spin_unlock(&kvm_ptp_lock);
++
++        return 0;
++}
++#endif
+ 
+ static int ptp_kvm_getcrosststamp(struct ptp_clock_info *ptp,
+-				  struct system_device_crosststamp *xtstamp)
++                                  struct system_device_crosststamp *xtstamp)
+ {
+-	return get_device_system_crosststamp(ptp_kvm_get_time_fn, NULL,
+-					     NULL, xtstamp);
++        return get_device_system_crosststamp(ptp_kvm_get_time_fn, NULL,
++                                             NULL, xtstamp);
+ }
+ 
+ /*
+@@ -105,64 +178,58 @@ static int ptp_kvm_getcrosststamp(struct ptp_clock_info *ptp,
+ 
+ static int ptp_kvm_adjfreq(struct ptp_clock_info *ptp, s32 ppb)
+ {
+-	return -EOPNOTSUPP;
++        return -EOPNOTSUPP;
```

Comment:
Is this and the below few changes required. Seem like just format changes.

Reply:
let me refine it @amshinde thanks.